### PR TITLE
FCP: Language bindings

### PR DESCRIPTION
**Pull request by mikelodder7**

### CODE REVIEW

**Comment by mikelodder7** [comment]
So I'm not opposed to using like an Array object or SizedBuffer but both sides of the interface must know about it. Rust has been interesting when you want to create arrays and pass them across the FFI boundary. In those cases you might have to create the struct in a separate call which makes it more complex and subject to more misuse. We can try to make a C like struct and see if its possible to use that across the boundary in a clean way.

File: text/language-bindings/README.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+stay in-sync with any other changes to the core library.
+
+Idiomatic wrappers usually cannot be generated by hand as they require more in depth language experience
+than automatic generators can produce. Therefore, idiomatic wrappers should be written by project contributors
+and maintained in separate repositories. These wrappers also SHOULD include tests for interacting with
+native Ursa.
+
+Here are general guidelines for producing wrappers:
+
+`<language>` should follow the naming convention as described [here](https://support.codebasehq.com/articles/tips-tricks/syntax-highlighting-in-markdown)
```
Comment (lovesh) [inline text/language-bindings/README.md]: Is the link appropriate? Where are the naming conventions?

**Comment by mikelodder7** [comment]
@dhuseby showed me [this](https://crates.io/crates/ffi-support) crate that I really like and matches patterns I've used in the past with additional safety guarantees. I'd recommend everyone to take a look at it.

File: text/language-bindings/README.md (hunk 1 above)
Comment (dcmiddle) [reply text/language-bindings/README.md]: Looks like this is marked as resolved but the link isn't changed?

**Comment by hartm** [comment]
I'm approving, as per our discussion last week.  I'll defer to the expertise of people on here--this kind of practical stuff is not my strong point.

File: text/language-bindings/README.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        - The "level" of the wrapper API can be generally "higher" than the C-callable functions, but it should expose the same amount of functionality and flexibility as the C-callable functions, so the higher level functions, if they exist, should be an addition to lower-level ones, not substitutes for them. In other words, helpers and convenience methods are great, but don’t allow them to entirely hide the core functions which provide granular control.
+    - The wrapper should document the earliest and latest version of ursa that it knows to be compatible.
+    - The wrapper should document what platforms it targets, what use cases it's built for, etc. A wrapper should be able to find ursa using default OS methods (e.g., in the system PATH), but should also provide a way for a specific path to ursa to be specified, such that the wrapper can work either from an OS-wide install of ursa or a version in a particular directory.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Some developers may want all ursa code written in the language of their needs to avoid issues presented by FFI.
```
Comment (vaporos) [inline text/language-bindings/README.md]: Maybe also mention how this relates to signing in-browser with Javascript. (Probably via wasm?)

**Comment by dhuseby** [comment]
I think to move forward we should rework this RFC to use the ffi-support crate and serialization/deserialization across the FFI boundary either with flatbuffers or capnproto or something like that. Protobufs has much wider support among the languages that would be calling Ursa (e.g. Python, Java, Javascript, Golang, etc) so maybe that's a much smarter direction to go in.

**Comment by mikelodder7** [comment]
@dhuseby feel free to fork my branch and submit a PR to mine then I’ll merge it with mine

File: text/language-bindings/README.md (hunk 1 above)
Comment (mikelodder7) [reply text/language-bindings/README.md]: At the link, you should see after the second code section `Here's a full list of supported languages:`. Following that is a list of abbreviated languages. That is what I am referring to.

File: text/language-bindings/README.md (hunk 2 above)
Comment (mikelodder7) [reply text/language-bindings/README.md]: So you want an example?

File: text/language-bindings/README.md (hunk 1 above)
Comment (dcmiddle) [reply text/language-bindings/README.md]: Ok, so this is the naming convention for paths/repos for languages, i.e. not java-script but js.

**Comment by mikelodder7** [comment]
> Ok, so this is the naming convention for paths/repos for languages, i.e. not java-script but js.

Yes

**Comment by vaporos** [comment]
Is the intent that "“libursa/bindings/<language>/folder” is checked into the repo or is the intent that it is generated as part of the build?

**Comment by mikelodder7** [comment]
> Is the intent that "“libursa/bindings//folder” is checked into the repo or is the intent that it is generated as part of the build?

Both actually. We would have libursa/bindings that would be checked into the repo and be as autogenerated as possible. Not all bindings can be. For now I recommend just doing C bindings as requests happen.

**Pull request merged**